src/libcore/pin.rs

```diff
@@ -7,23 +7,31 @@
 //! since moving an object with pointers to itself will invalidate them,
 //! which could cause undefined behavior.
 //!
-//! In order to prevent objects from moving, they must be pinned
-//! by wrapping a pointer to the data in the [`Pin`] type. A pointer wrapped
-//! in a `Pin` is otherwise equivalent to its normal version, e.g. `Pin<Box<T>>`
-//! and `Box<T>` work the same way except that the first is pinning the value
-//! of `T` in place.
+//! By default, all types in Rust are movable. Rust allows passing all types by-value,
+//! and common (smart-)pointer types such as `Box`, `Rc`, and `&mut` allow replacing and
+//! moving the values they contain. The [`Pin`] type provides a wrapper for these pointer
+//! types which prohibits moving the value behind the pointer. For example, `Pin<Box<T>>`
+//! functions much like a regular `Box<T>`, but doesn't allow moving `T`. The pointer
+//! value itself (the `Box`) can still be moved, but the value behind it cannot.
 //!
-//! First of all, these are pointer types because pinned data mustn't be passed around by value
-//! (that would change its location in memory).
-//! Secondly, since data can be moved out of `&mut` and `Box` with functions such as [`swap`],
-//! which causes their contents to swap places in memory,
-//! we need dedicated types that prohibit such operations.
+//! Since data can be moved out of `&mut` and `Box` with functions such as [`swap`],
+//! changing the location of the underlying data, [`Pin`] prohibits accessing the
+//! underlying pointer type (the `&mut` or `Box`) directly, and provides its own set of
+//! APIs for accessing and using the value. [`Pin`] also guarantees that no other
+//! functions will move the pointed-to value. This allows for the creation of
```

Comment:
I think we may want to say "no later code", or something along those lines, instead of "other functions". The contract around `Pin` talks about what cannot happen "in the future", but the current text doesn't currently say that.